File: 68. Text Justification/fullJustify.py
from typing import List

class Solution:
  def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
    res = []
    line, lineLength = [], 0
    i = 0

    while i < len(words):
      if lineLength + len(line) + len(words[i]) > maxWidth:
        extraSpaces = maxWidth - lineLength
        spaces = extraSpaces // max(1, len(line) - 1)
        reminderSpaces = extraSpaces % max(1, len(line) - 1)

        for j in range(max(1, len(line) - 1)):
          line[j] += " " * spaces
          if reminderSpaces:
            line[j] += " "
            reminderSpaces -= 1
        
        res.append(''.join(line))
        line, lineLength = [], 0


      line.append(words[i])
      lineLength += len(words[i])
      i += 1
    
    lastLine = ' '.join(line)
    trailingSpaces = maxWidth - len(lastLine)
    res.append(lastLine + " " * trailingSpaces)

    return res
        

words = ["This", "is", "an", "example", "of", "text", "justification."]
maxWidth = 16
solution = Solution()
print('fullJustify: ', solution.fullJustify(words, maxWidth))

File: 68. Text Justification/fullJustify.js
/**
 * @param {string[]} words
 * @param {number} maxWidth
 * @return {string[]}
 */
const fullJustify = (words, maxWidth) => {
  const res = [];
  let line = [];
  let lineLength = 0;
  let i = 0;

  while (i < words.length) {
    if (lineLength + line.length + words[i].length > maxWidth) {
      const extraSpaces = maxWidth - lineLength;
      const spaces = Math.floor(extraSpaces /  Math.max(1, line.length - 1));
      let reminderSpaces = extraSpaces % Math.max(1, line.length - 1)

      for (let j = 0; j < Math.max(1, line.length - 1); j++) {
        line[j] += " ".repeat(spaces)
        if (reminderSpaces) {
          line[j] += " ";
          reminderSpaces--;
        }
      }

      res.push(line.join(''));
      line = [];
      lineLength = 0;
    }

    line.push(words[i]);
    lineLength += words[i].length;
    i++;
  }

  const lastLine = line.join(' ');
  const trailingSpaces = maxWidth - lastLine.length;
  res.push(lastLine + " ".repeat(trailingSpaces));

  return res;
};


const words = ["This", "is", "an", "example", "of", "text", "justification."];
const maxWidth = 16;
console.log('fullJustify: ', fullJustify(words, maxWidth));
